### DISCUSSION TITLE
How to configure global rate limiting in `ocelot.json` file?

### DISCUSSION THREAD

**Discussion by perisicnikola37** [ThreeMammals/Ocelot#2164] [Q&A] [answered]
How to enable global rate limiting on all defined endpoints without repeating the definitions at the end of each endpoint?

This doesn't work:

```bash
{
  "Routes": [
    {
      "DownstreamPathTemplate": "/api/accounts",
      "DownstreamScheme": "http",
      "DownstreamHostAndPorts": [
        {
          "Host": "account_api",
          "Port": "8080"
        }
      ],
      "UpstreamPathTemplate": "/accounts",
      "UpstreamHttpMethod": ["POST"]
    },
  ],
  "GlobalConfiguration": {
    "BaseUrl": "http://localhost:8010",
    "RateLimitOptions": {
      "ClientWhitelist": [],
      "EnableRateLimiting": true,
      "Period": "10s",
      "PeriodTimespan": 10,
      "Limit": 5
    }
  }
}

```

**Comment by raman-m** [comment, marked as answer]
Hello, Nikola! And welcome to the world of Ocelot! 🐯 

You are not alone, and such a question regarding global Rate Limiting options was asked by other developers:
- #1229 
- https://github.com/ThreeMammals/Ocelot/issues/1915#issuecomment-1885372796
- https://github.com/ThreeMammals/Ocelot/issues/585#issuecomment-1832711326 
  This discussion clarifies why the feature has not been implemented yet, as applying the same rate limiting to all static routes is considered poor DevOps practice.

To set up dynamic routing and apply global settings, consider reading the following documentation:
- [Rate Limiting | Global Configuration](https://github.com/ThreeMammals/Ocelot/blob/develop/docs/features/ratelimiting.rst#global-configuration)
- [Dynamic Routing](https://github.com/ThreeMammals/Ocelot/blob/develop/docs/features/servicediscovery.rst#dynamic-routing:~:text=RateLimitOptions): So, global rate limiting opts are available in dynamic routing mode. You have to setup service discovery: Consul, K8s, Eureka etc.

**Reply by raman-m** [reply]
Nikola, I've reopened both issues #1229 and #1915, and we aim to deliver this feature by [November'24](https://github.com/ThreeMammals/Ocelot/milestone/9). Is that 🆗?

Your contributions, including any PR, would be greatly appreciated, especially now that you seem to be coding in C# 😉

**Reply by perisicnikola37** [reply]
Appreciate you reopening the issues. November sounds fine to me 🙂. 

As for contributing, I might drop something I'll see 🙂

**Reply by raman-m** [reply]
Also, after [blaming the doc](https://github.com/ThreeMammals/Ocelot/blame/develop/docs/features/ratelimiting.rst#global-configuration) I've realized we forgot to mention (in commit https://github.com/ThreeMammals/Ocelot/commit/aef3e6b9f53ad8c8aa8c9ade92a572f60d0672a1)  that Global Configuration is available in Dynamic Routing mode only... This may cause confusion for you and other developers.
So, I'm going to update docs soon...

**Reply by raman-m** [reply]
Documentation was updated in release [23.3.5](https://github.com/ThreeMammals/Ocelot/releases/tag/23.3.5). Here is the updated [Rate Limiting | Global Configuration](https://ocelot.readthedocs.io/en/latest/features/ratelimiting.html#global-configuration) docs.

**Reply by perisicnikola37** [reply]
Thank you

**Reply by perisicnikola37** [reply]
@raman-m Hi. Is this feature available now?

**Reply by raman-m** [reply]
@perisicnikola37 No! Unfortunately, I'm swamped with the [.NET 9](https://github.com/ThreeMammals/Ocelot/milestone/11) release. Our team capacity is currently limited as other members are not actively working. Please reach out to me again after the release, as prioritization can only be addressed post the .NET 9 version.

**Reply by raman-m** [reply]
**One question**: Do you actually use header-based limiting, also known as [Rate Limit by Client's Header](https://ocelot.readthedocs.io/en/latest/features/ratelimiting.html#ocelot-own-implementation)? This feature in Ocelot is poorly designed and has flaws. Have you considered the [ASP.NET Implementation](https://ocelot.readthedocs.io/en/latest/features/ratelimiting.html#future-and-asp-net-implementation), which is a native Microsoft feature?
